Style Book: Move iframe to root of content area to support styles that overflow block previews

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Fixes: https://github.com/WordPress/gutenberg/issues/48392

Alternative to #48470, this PR looks at moving the iframe for the Style Book up to the root of the content area, and uses editor context / block list at the example level, so that the styles affect and are displayed across the whole content area of the Style Book rather than just within each particular block example.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Styles need to affect an area beyond just their preview — the idea in this PR is that (hopefully) we can get the styles to affect the entire content area so that it more closely resembles the layout of the editor.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

* Add an iframe at the root of the content area of the style book
* Add some hard-coded styles to pass to that iframe along with editor styles
* At the individual example level, use a block editor context provider and block list for rendering the block previews
* Reset button styles and then include hard-coded rules for the button element, copying/pasting rules from the Button component as needed (e.g. focus styles)

Note / caveat: because of lifting the iframe up to the root of the content area, I had to set the iframe to `tabIndex=0` so that the buttons within it could be tabbed to. This seems to work pretty well, but means that the iframe itself now sits within the tab sequence of the style book. Happy for any suggestions on how to improve this.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. Open the Style Book and select Button under Design.
2. Adjust the Shadow settings for the Button block, and note that with this PR applied, the bigger shadows should still be visible outside of the area of the block example.
3. Double-check that other styles are appearing correctly, too, e.g. set the Image block to have a Duotone value set. The Duotone styling should be visible in the Style Book, too.
4. Try with a variety of themes with custom fonts, etc — check that the styling looks correct in the Style Book / matches `trunk`, and that the block headings still match the styling in Gutenberg

For code review: double-check whether or not I've included all the bits required from the block previews over to the examples here.

### Testing Instructions for Keyboard

1. Open the Style Book within global styles.
2. From this point tab to the examples within the Style Book content area. You should be able to tab between / press Enter to select any of the previews as on trunk. The main difference is that the iframe container area is now in the tab sequence. Again, happy for any ideas on how to improve this / skip the iframe in the tab sequence!

## Screenshots or screencast <!-- if applicable -->

Button block with a large shadow applied in a site with a gradient set as the site's background color:

![image](https://user-images.githubusercontent.com/14988353/223301203-49b6db36-f0a0-4bd7-b160-c8d6a8d6c6e2.png)

andrewserong committed:
Style Book: Try moving iframe to root of content area

andrewserong committed:
Try to match styles inside the iframe to the existing text styles from Gutenberg

andrewserong committed:
Restore margin rules to prevent vertical margins on previews from affecting the height of the preview

andrewserong committed:
Fix svg filters, attempt to fix tabbing behaviour

andrewserong committed:
Add Enter key handling, button role, update e2e tests

andrewserong committed:
Tidy hard-coded styling rules

andrewserong commented:
@joedolson just adding you as a reviewer for this one, too, as we've been looking at accessibility aspects of the Style Book lately. This PR seeks to swap the `button` for a `div` due to dealing with styling of the block preview that appears within the button (so that we can avoid button styles bleeding into / affecting the style previews). I _think_ I've mostly got the button behaviour working now, but keen for any feedback if there are improvements to be made here 🙂

kevin940726 commented on test/e2e/specs/site-editor/style-book.spec.js:
```diff
@@ -59,37 +59,50 @@ test.describe( 'Style Book', () => {
 
+		// Buttons to select block examples are rendered within the Style Book iframe.
+		let styleBookIframeLocator = page.frameLocator(
+			'.edit-site-style-book__iframe'
```
Is there a `name` attribute for the frame? Maybe it should? If so we can just do `page.frameLocator( '[name="name"]' )` or `page.frame( name )` here.

andrewserong replied:
The `page.frameLocator( '[name="name"]' )` approach worked nicely here, I've added a `style-book-canvas` name to the iframe. When testing locally, I noticed that `page.frameLocator()` appears to support reevaluating the iframe as needed, however `page.frame()` appeared to grab a specific instance of a frame, so failed after we switch tabs. I've gone with `page.frameLocator` in both tests for consistency, but targeted by name.

kevin940726 commented on test/e2e/specs/site-editor/style-book.spec.js:
```diff
@@ -59,37 +59,50 @@ test.describe( 'Style Book', () => {
 		await page.click( 'role=tab[name="Media"i]' );
 
+		// The tab has changed, which means the iframe has been re-rendered.
+		styleBookIframeLocator = page.frameLocator(
```
If we're using `frameLocator` then we don't need to reevaluate since they will be evaluated lazily anyway. We can reuse the same locator.

andrewserong replied:
Oh, TIL, thanks Kai!

andrewserong replied:
This worked nicely!

tellthemachines commented:
A quick test shows this working well to solve the box-shadow problem! It's also loading the stylebook noticeably faster than on trunk 🎉 

> Swap the button element used for a div to avoid button styling from affecting the preview

Is it just the default UA styles we're trying to avoid? I wonder if it would be easier to do something like `all: unset` on the button element instead. 

If we do go the div route, we'll also need a handler for the Space key.

I'll do a proper review next week, but this is definitely looking like the way to go!

jasmussen commented:
Thanks for tackling this one, it feels like a much better solution. 🙏 

From the visuals it looks right.

apeatling reviewed (commented):
This is working really well in my testing. Button shadows now looking exactly as expected. I also confirmed that the duotone filter is working as expected on images:

![2023-03-03 09 28 13](https://user-images.githubusercontent.com/1464705/222789467-6ff5937e-88d1-4dd8-bdb4-76f4a7d38600.gif)

I tested two themes with more interesting fonts, Sassify and Tortoise. Both showed their fonts in the style book as expected and I saw no difference between trunk and this branch.

Sassify
<img width="1139" alt="sassify" src="https://user-images.githubusercontent.com/1464705/222789759-c8161057-d021-44eb-81b4-f14c83bc5ad6.png">

Tortoise
<img width="1427" alt="Screenshot 2023-03-03 at 9 32 26 AM" src="https://user-images.githubusercontent.com/1464705/222789792-39c68cd1-e032-4a69-ba85-eafec56e5452.png">

I didn't check the code, but looks like @kevin940726 is giving that a look over.

joedolson commented:
The `div` replacement for a button mostly works, though the spacebar isn't handled as mentioned above. I'll also note that space and enter shouldn't be handled the same way - `space` should be handled on keyup, and `enter` on keydown to match native behavior. 

I'm not totally clear on why this was necessary, however; why were button styles bleeding between those blocks and the button? The button block creates an `a href=""` pattern, not a button, so I don't understand why that would bleed.

I'm reluctant to embrace using a faux button unless we really have to, and right now I don't see the reasoning. It'll need to be thoroughly tested on all standard screen reader/browser/os combinations, since there can be a lot of variation in how they handle custom events.

andrewserong commented:
Thanks for the feedback, all!

> I'm not totally clear on why this was necessary, however; why were button styles bleeding between those blocks and the button? The button block creates an a href="" pattern, not a button, so I don't understand why that would bleed.

Previous to this PR, this one was actually using a simple `button` element, and not the `Button` component that renders out an `a` tag. Because there used to be an iframe inside the button, there was less risk of styles bleeding into the preview. Now that there isn't another iframe, it's a bit more important to make sure that there aren't any styles flowing through.

> I wonder if it would be easier to do something like all: unset on the button element instead.

I gave `all: unset` a try, but it didn't work for me in testing. I might give that another go.

From the sounds of it we're heading in a good direction, thanks for the encouragement everybody! For next steps, I'll try using a `button` element again, and if I can't get the styles to play nicely, I'll give a simple `a` element a try instead of making the `div` interactive 👍

andrewserong committed:
Switch back to button element, add additional focus styles, update e2e tests

andrewserong commented:
Update: the suggestion from @tellthemachines of using `all: unset` worked nicely! I've updated the CSS so it sets that, and then there's another couple of rules for applying the styling that we do want to include for the buttons. I copied over the CSS from the Button components to provide focus styles. It's testing well for me locally now across Safari, FF, Chrome, and Edge, but let me know if anyone runs into any issues with it. Overall, this feels much nicer than trying to turn a `div` into a button 👍

I've also updated the e2e tests based on the suggestions from @kevin940726, so this PR should be ready for another review now 🙂

tellthemachines commented on packages/edit-site/src/components/style-book/index.js:
```diff
@@ -38,12 +43,86 @@ import { ESCAPE } from '@wordpress/keycodes';
+		background: none;
+		border-radius: 2px;
+		border: none;
+		color: inherit;
```
We shouldn't need to set background and border to `none`, or color to `inherit` here because we've already unset everything above.

andrewserong replied:
Oh, good catch, I'll tighten these up 👍

tellthemachines reviewed (approved):
Code's looking pretty good to me, aside from one minor comment below!

Testing on Chrome, Firefox and Safari on macOS didn't reveal any issues ✅ 

It's also pretty exciting that theme gradient and pattern backgrounds now work seamlessly! E.g. this is the Pilgrimage TT3 variation on trunk:
<img width="951" alt="Screenshot 2023-03-06 at 4 37 48 pm" src="https://user-images.githubusercontent.com/8096000/223027579-aa91f14c-728e-45b4-917e-e7efa616d004.png">


vs this branch:

<img width="873" alt="Screenshot 2023-03-06 at 4 37 28 pm" src="https://user-images.githubusercontent.com/8096000/223027614-21dab699-3cc1-40e6-bc5a-1516d57e1269.png">

Note how on trunk the dots only appear in a little window around each block. It's looking much better with this change!

kevin940726 commented on test/e2e/specs/site-editor/style-book.spec.js:
```diff
@@ -59,37 +59,45 @@ test.describe( 'Style Book', () => {
-			'role=button[name="Open Headings styles in Styles panel"i]'
-		);
+		await page
+			.frameLocator( '.edit-site-style-book__iframe' )
```
Missed this? 😆

andrewserong replied:
I was too excited to commit, I totally missed this one, thanks for the eagle eyes! 👀 😀

kevin940726 replied:
There's one more below too 😆

andrewserong replied:
Hang on, isn't that one already updated?

kevin940726 replied:
Oh, oops! Right! GitHub failed me somehow 😅 .

andrewserong replied:
Thanks for confirming — I've been staring at this diff for too long, so I thought I very well could have missed something 😄

andrewserong committed:
Remove unneeded CSS rules

andrewserong committed:
Update e2e test

andrewserong commented:
Thanks again @tellthemachines and @kevin940726, I've removed the unneeded button styles resets already covered by `all: unset` and updated the e2e test I missed 😄

kevin940726 reviewed (commented):
I didn't look into the code, but the testing changes look good to me 👍 .

andrewserong commented:
I'm wrapping up for the day now, but just added the backport to WP Beta/RC label in case folks are hoping to get this in for the next beta/RC.

ntsekouras commented on packages/edit-site/src/components/style-book/index.js:
```diff
@@ -156,65 +241,105 @@ function StyleBook( { isSelected, onSelect, onClose } ) {
+	);
+
+	return (
+		<button
```
This can lead to invalid HTML(`<button> cannot appear as a descendant of <button>`).  Can we avoid this by using `Composite`?

tellthemachines replied:
That's a good point, I don't think we have any actual buttons in core blocks, but we do have other interactive elements. 

I wonder if we could wrap the whole thing in a div, and use a button to wrap the block name, and then have the actual block sit beside it. It might be trickier to activate the outline when the button is clicked but we may be able to leverage `:has()` for that?

andrewserong replied:
Thanks for the feedback! I'll have a go at using `Composite`, I see the [block patterns list uses `Composite`](https://github.com/WordPress/gutenberg/blob/94ff2ba55379d9ad7f6bed743b20b85ff26cf56d/packages/block-editor/src/components/block-patterns-list/index.js#L130) and `CompositeItem` so it looks like we'll have some consistency there if we switch to that.

>  I don't think we have any actual buttons in core blocks, but we do have other interactive elements.

I tried installing a couple of plugins and I noticed that WooCommerce includes buttons in their cart block, so I imagine it's fairly likely that we'll encounter more 3rd party blocks that use `button` elements.

I'll have a play and see what our options are 🙂

tellthemachines replied:
Note that whatever we go with here will still be an improvement over trunk, where we have a whole iframe inside each button 😅

andrewserong replied:
I've had a go with Composite and CompositeItem in https://github.com/WordPress/gutenberg/pull/48664/commits/cf0aec6de0afc5c7914f4d72a32e564e5387a6cc, seems to be working pretty well so far!

madhusudhand reviewed (approved):
This tests very well! 
Verified with various themes and shadow definitions. It doesn't show any issues. 🚢

andrewserong committed:
Improve caching of blocks

andrewserong committed:
Try swapping out button element for Composite and CompositeItem

andrewserong commented:
Update: based on the suggestion from @ntsekouras (https://github.com/WordPress/gutenberg/pull/48664#discussion_r1126057655) I've swapped out the `button` element again, this time using `Composite` and `CompositeItem` which are internally using [`Composite` from reakit](https://reakit.io/docs/composite/). I've followed the performance suggestion in the docs there to add an `id` attribute on the `CompositeItem`, too.

This appears to give us the best of both worlds in that we avoid the invalid HTML of rendering buttons within buttons (since 3rd party blocks are likely to contain them), while also getting the keyboard behaviour from `Composite` which allows using up/down keys to move through the list of previews. From manually testing, it also appears to come with the right enter/space key behaviour as well. There's also some consistency here in that the [block patterns list uses `Composite` here](https://github.com/WordPress/gutenberg/blob/94ff2ba55379d9ad7f6bed743b20b85ff26cf56d/packages/block-editor/src/components/block-patterns-list/index.js#L130).

I'm not too sure if I've gotten the `role` values right on `Composite` / `CompositeItem` — for the latter I've left it as `button`, but I see that other usage of `Composite` in Gutenberg uses `listbox`, which doesn't _quite_ feel right for this use case, as users aren't really selecting an option from a list? Happy for feedback / ideas on what makes most sense here! CC: @joedolson as you mentioned concerns with the accessibility of using `div` elements that have the behaviour of buttons.

On balance, I _think_ `Composite` sounds preferable to either a simple `button` element (since it doesn't support the nested content we need for previews) or a `div` element (since we'd have to write our own handling to make sure it behaves correctly).

tellthemachines commented on packages/edit-site/src/components/style-book/index.js:
```diff
@@ -156,65 +245,133 @@ function StyleBook( { isSelected, onSelect, onClose } ) {
+				className={ classnames( 'edit-site-style-book__example', {
+					'is-selected': isSelected,
+				} ) }
+				id={ id }
```
Is the id being used for anything?

andrewserong replied:
Yep! It's a little weird, but I followed the suggestions from [the reakit docs here](https://reakit.io/docs/composite/#performance) that recommend it as a performance enhancement when using `CompositeItem` components when you're concerned about performance. I figured since we're dealing with previews and we want to avoid re-renders where we can, it'd be a good one to include. It's the first time I've used it, though, so very happy for feedback if I've missed anything 😄

tellthemachines reviewed (commented):
It's still working well for me! 

A quick test with VoiceOver yielded the same results as trunk, but it might be good to test on Windows too if anyone has the setup for it. What I'm unsure about is to what extent having a div with a button role is any better than having an actual button, especially as regards its possible interactive content. I strongly suspect none of these changes will make things any worse than they already are on trunk though 😅

Mamaduka commented on packages/edit-site/src/components/style-book/index.js:
```diff
@@ -156,65 +245,133 @@ function StyleBook( { isSelected, onSelect, onClose } ) {
+					.map( ( example, index ) => (
+						<Example
+							key={ example.name }
+							id={ `example-${ index }` }
```
Nit: Maybe we should have the `id` unique like a `key` - `example-${ example.name }`? What do you think?

andrewserong replied:
Thanks! I was mostly just following the performance example from the reakit docs, but this feels neater and more consistent with how we do things in Gutenberg (using block name as a unique key). Updated in https://github.com/WordPress/gutenberg/pull/48664/commits/f1a2c35466ba8e2f3fad5d67e958878fffca12b4

Mamaduka commented on packages/edit-site/src/components/style-book/index.js:
```diff
@@ -156,65 +245,133 @@ function StyleBook( { isSelected, onSelect, onClose } ) {
+			[ originalSettings ]
+		);
+
+		// Cache the list of blocks to avoid additional processing when the component is re-rendered.
+		const renderedBlocks = useMemo(
+			() => ( Array.isArray( blocks ) ? blocks : [ blocks ] ),
+			[ blocks ]
+		);
```
If the `blocks` prop is unstable, wrapper `memo` becomes useless.

andrewserong replied:
Oh, good point! I noticed yesterday that the casting to an array logic was causing additional re-renders / processing, and this `useMemo` was required to stabilise `blocks`. From re-testing, it looks like the wrapper `memo` is no longer needed as you mention, so I've removed it in https://github.com/WordPress/gutenberg/pull/48664/commits/f1a2c35466ba8e2f3fad5d67e958878fffca12b4 — on `trunk` it seems like `blocks` was the reason this component was wrapped in a `memo` call, so the level of caching now should be basically the same as before I think.

andrewserong committed:
Update id key, remove unnecessary memo call

andrewserong commented:
I've done a couple of tiny updates, thanks to suggestions from @Mamaduka. Also, I've included an `aria-hidden` on the preview `div`, as that was one piece of the block preview component I'd forgotten to add in.

I think this is in a good shape to land now, and it's an improvement over what we have in trunk as @tellthemachines mentions. If there aren't any objections, I'll aim to merge this PR in tomorrow my time, but feel free to merge @Mamaduka or @ntsekouras if this looks okay and if it makes the backporting process easier to have it in sooner.

Thanks again for all the reviews and feedback, everybody! 🙇

ntsekouras reviewed (approved):
Thank you @andrewserong, this looks good! It would be good to see if we can find a better solution for the extra iframe tab stop, but for now this is an improvement against trunk.

Mamaduka commented:
I just cherry-picked this PR to the wp/6.2 branch to get it included in the next release: 3082500a53
